shopfloor_reception/services/reception.py

```diff
@@ -248,27 +245,37 @@ def _select_document_from_lot(self, lot):
         )
 
     def _scan_line__find_or_create_line(self, picking, move, qty_done=1):
-        line = fields.first(
-            move.move_line_ids.filtered(
-                lambda l: (
-                    not l.result_package_id
-                    and l.shopfloor_user_id.id in [False, self.env.uid]
-                )
-            )
-        )
-        if line:
-            # The line quantity to do needs to correspond to
-            # the remaining quantity to do of its move.
-            line.product_uom_qty = move.product_uom_qty - move.quantity_done
-        else:
-            qty_todo_remaining = max(0, move.product_uom_qty - move.quantity_done)
-            values = move._prepare_move_line_vals(quantity=qty_todo_remaining)
+        """Find or create a line  on a move for the user to work on.
+
+        First try to find a line already assigned to the user.
+        Then a line that is not yet assigned to any users (locking the line
+        to avoid concurent access.)
+        If none are found create a new line.
+
+        """
+        line = None
+        unassigned_lines = self.env["stock.move.line"]
+        for move_line in move.move_line_ids:
+            if move_line.result_package_id:
+                continue
+            if move_line.shopfloor_user_id.id == self.env.uid:
+                line = move_line
+                break
+            elif not move_line.shopfloor_user_id:
+                unassigned_lines |= move_line
+        if not line and unassigned_lines:
+            lock = self._actions_for("lock")
+            for move_line in unassigned_lines:
+                if lock.for_update(move_line, skip_locked=True):
+                    line = move_line
+                    break
+        if not line:
+            values = move._prepare_move_line_vals()
             line = self.env["stock.move.line"].create(values)
```

Comment:
I think we should ensure that the product_uom_qty is always 0. 

```suggestion
        if not line:
            values = move._prepare_move_line_vals()
            line = self.env["stock.move.line"].create(values)
        else:
            line.product_uom_qty = 0
```

Reply:
We do this on the post for backward compatibility with previous implementation (lines that would already be there with a qty != 0). Also if you have only one line pre-generated by odoo, you have a qty != 0 and this is not causing any issue.
When posting, it is required we care other lines (if any) are well set to 0 otherwise you cannot post.

Reply:
https://github.com/OCA/wms/pull/726#discussion_r1394116955